### PR TITLE
Feature/batchtx

### PR DESCRIPTION
**Pull request by ed255**
# Close Database on failure

# Batch transactions in the StateDB

Introduce db/batch.Batch: a wrapper over a db.WriteTx that implements
db.WriteTx.  The batch acts as a WriteTx but when it becomes too big, it
automatically commits and continues with a new internal db.WriteTx
transparently.

Use this Batch instead of a regular WriteTx in the statedb.TreeTx to
avoid hitting the db.WriteTx size limit when the block contains many
updates.  Since we are now writting to disk between StateDB versions,
when opening TreeUpdates we need to make sure we start always from the
last commited version, so we query the expected root and set it via
tree.SetRoot if necessary.

This should fix the observed error "Txn is too big to fit into one
request" seen in #280

### CODE REVIEW

File: censustree/censustree.go [hunk 1]
```diff
@@ -51,6 +51,11 @@ func New(wTx db.WriteTx, opts Options) (
 	if err != nil {
 		return nil, err
 	}
+	defer func() {
+		if err != nil {
```
Comment (mvdan) [inline censustree/censustree.go]: what err are you reading here? supposedly the one from tree.New, but that's a `:=`, so you're actually shadowing.

IMO it's much better if you avoid a named return parameter and defer func, and just call `database.Close` in the only case that matters: when `tree.New` fails.

File: censustreelegacy/arbotree/wrapper.go [hunk 2]
```diff
@@ -33,12 +33,17 @@ var _ censustree.Tree = (*Tree)(nil)
...
-	censustree.Tree, error) {
+	_ censustree.Tree, err error) {
 	dbDir := filepath.Join(storageDir, "arbotree.db."+strings.TrimSpace(name))
 	database, err := badgerdb.New(badgerdb.Options{Path: dbDir})
 	if err != nil {
 		return nil, err
 	}
+	defer func() {
+		if err != nil {
+			database.Close()
```
Comment (mvdan) [inline censustreelegacy/arbotree/wrapper.go]: same as before.

File: db/badgerdb/badger.go [hunk 3]
```diff
@@ -53,12 +53,20 @@ func (tx WriteTx) Get(k []byte) ([]byte, error) {
 
 // Set implements the db.WriteTx.Set interface method
 func (tx WriteTx) Set(k, v []byte) error {
-	return tx.tx.Set(k, v)
+	if err := tx.tx.Set(k, v); err == badger.ErrTxnTooBig {
```
Comment (mvdan) [inline db/badgerdb/badger.go]: use `errors.Is` rather than `==`

File: tree/tree.go [hunk 4]
```diff
@@ -227,6 +227,24 @@ func (t *Tree) FromRoot(root []byte) (*Tree, error) {
 	}, nil
 }
 
+func (t *Tree) SetRoot(wTx db.WriteTx, root []byte) error {
+	givenTx := true
+	if wTx == nil {
+		givenTx = false
```
Comment (mvdan) [inline tree/tree.go]: nit: easier if you do

```
givenTx := wTx != nil
if !givenTx {
```

this way it also mirrors the other `if !givenTx` check, so it's easier to understand they are matching each other

File: vochain/state.go [hunk 5]
```diff
@@ -634,14 +634,15 @@ func (v *State) Envelope(processID, nullifier []byte, isQuery bool) (_ []byte, e
...
 func (v *State) EnvelopeExists(processID, nullifier []byte, isQuery bool) (bool, error) {
-	e, err := v.Envelope(processID, nullifier, isQuery)
-	if err != nil && err != ErrVoteDoesNotExist {
-		return false, err
-	}
-	if err == ErrVoteDoesNotExist {
+	_, err := v.Envelope(processID, nullifier, isQuery)
+	if err == ErrProcessNotFound {
+		return false, nil
+	} else if err == ErrVoteDoesNotExist {
```
Comment (mvdan) [inline vochain/state.go]: `errors.Is` for these too; this helps ensure that future error wrapping won't break our code.

File: db/batch/batch.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
+package batch
+
+import "go.vocdoni.io/dvote/db"
+
+// Batch is a wrapper over a db.WriteTx that implements db.WriteTx.  The batch
+// acts as a WriteTx but when it becomes too big, it automatically commits and
+// continues with a new internal db.WriteTx transparently.
```
Comment (mvdan) [inline db/batch/batch.go]: > Batch is a wrapper over a db.WriteTx that implements db.WriteTx

That feels a bit redundant :P How about just keeping the second part, with a small modification at the start:

```
// Batch wraps a WriteTx to automatically commit when it becomes too big,
// causing a new internal WriteTx to be created.
```

File: db/batch/batch.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
+package batch
```
Comment (mvdan) [inline db/batch/batch.go]: why make this a new package? tiny packages are usually not a good idea, as they just make your API harder to use and the code harder to develop. Plus, `db.Batch` is IMO a much nicer name than `batch.Batch` - less repetitive, and it clarifies what its purpose is and where it fits.

File: db/batch/batch.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+func (t *Batch) Discard() {
+	t.tx.Discard()
+}
+
+// Set implements the db.WriteTx.Set interface method.  If during this
+// operation, the internal tx becomes too big, all the pending writes will be
+// committed and a new db.WriteTx will be created to continue with this and
+// future writes.
+func (t *Batch) Set(key []byte, value []byte) error {
+	if err := t.tx.Set(key, value); err == db.ErrTxnTooBig {
```
Comment (mvdan) [inline db/batch/batch.go]: `errors.Is` here too and any other places I might have missed

**Review by mvdan — APPROVED** [review]
Very nice tests!

**Review by arnaucube — APPROVED** [review]

File: censustree/censustree.go [hunk 9]
```diff
@@ -34,7 +34,7 @@ const nLevels = 256
 // New returns a new Tree, if there already is a Tree in the
 // database, it will load it.
 func New(wTx db.WriteTx, opts Options) (
-	*Tree, error) {
+	_ *Tree, err error) {
```
Comment (mvdan) [inline censustree/censustree.go]: probably undo these changes now :) also no need to split these func decls into two lines, they're fairly short

**Review by mvdan — APPROVED** [review]

File: db/batch/batch.go (hunk 7 above)
Comment (ed255) [reply db/batch/batch.go]: I've moved Batch to the db package.  In `batch_test.go` I was using the badgerdb implementation, so having the batch tests in the db package created an import cycle, so I moved `batch_test.go` to the badgerdb package (I think it makes some sense, as it's testing the Batch over badgerdb).

Comment (mvdan) [reply db/batch/batch.go]: That makes sense. Another possibility is to write tests in a `db_test` package, so then you can have db_test import badgerdb, and badgerdb importing db - no cycle.
